src/openvpn/tun2tap.c

```diff
@@ -0,0 +1,196 @@
+/*
+ *  OpenVPN -- An application to securely tunnel IP networks
+ *             over a single TCP/UDP port, with support for SSL/TLS-based
+ *             session authentication and key exchange,
+ *             packet encryption, packet authentication, and
+ *             packet compression.
+ *
+ *  Copyright (C) 2002-2018 OpenVPN Inc <sales@openvpn.net>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2
+ *  as published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+/*
+Author: pengtianabc@hotmail.com
+Date: 2020-4-12 11:17:33
+Function: simulate arp logical for tun device on client
+TODO: ship for server, but i think that not worth
+*/
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#elif defined(_MSC_VER)
+#include "config-msvc.h"
+#endif
+
+#include "syshead.h"
+// #include "openvpn.h"
+#include "tun2tap.h"
+
+#include "memdbg.h"
+
+static inline void dump_hex(const uint8_t *buf, int len, char *name){
+        int i = 0;
+        printf("===buf name:%s, len:%d==\n", name, len);
+        for(; i < len; i++){
+                printf("%02x", buf[i]);
+        }
+        printf("\n======================\n");
+}
+
+/*
+ * Should we handle arp to the remote?
+ */
+bool
+check_tun2tap_send_dowork(struct context *c, int flag)
+{
+    int ret = true;
+    if (!c->options.tun2tap){
+        return ret;
+    }
+    if (TUN2TAP_FLAG_ENCAP == flag){
+        dmsg(D_TUN2TAP, "TUN2TAP: encap data from tun");
+        // read from tun, check and send arp to link
+        if ( BLEN(&c->c2.buf) >= sizeof(struct openvpn_iphdr)){
+            struct openvpn_ethhdr hdr = {0};
+            struct openvpn_iphdr *ip_hdr = BPTR(&c->c2.buf);
+            struct openvpn_ipv6hdr *ipv6_hdr = BPTR(&c->c2.buf);
+            int v = OPENVPN_IPH_GET_VER(ip_hdr->version_len);
+            memcpy(hdr.dest, c->c1.tuntap->remote_mac_addr, OPENVPN_ETH_ALEN);
+            memcpy(hdr.source, c->options.lladdr_v, OPENVPN_ETH_ALEN);
+            
+            if (0 == memcmp(hdr.dest, "\x00\x00\x00\x00\x00\x00", sizeof("\x00\x00\x00\x00\x00\x00") - 1)){
```

Comment:
Write a comment *why* we are checking here against the all zeo mac. It is not really obvious